File: abelkhan_type_component/component/dbproxy/hub_msg_handle.ts
import abelkhan = require('../../abelkhan_type/ts/abelkhan');
import mongodbproxy = require('./mongodbproxy');
import hubmanager = require('./hubmanager');
import dbproxy_protcol = require('../../protcol/ts/dbproxy');
import log = require('../../service/log/log');

export class hub_msg_handle{
    private _hubmng : hubmanager.hubmanager;
    private _mongodbproxy : mongodbproxy.mongodbproxy;
    private hub_call_dbproxy_module : dbproxy_protcol.hub_call_dbproxy_module;
    constructor(modules:abelkhan.modulemng, _hubs:hubmanager.hubmanager, _dbproxy:mongodbproxy.mongodbproxy){
        this._hubmng = _hubs;
        this._mongodbproxy = _dbproxy;

        this.hub_call_dbproxy_module = new dbproxy_protcol.hub_call_dbproxy_module(modules);
        this.hub_call_dbproxy_module.cb_reg_hub = this.reg_hub.bind(this);
        this.hub_call_dbproxy_module.cb_create_persisted_object = this.create_persisted_object.bind(this);
        this.hub_call_dbproxy_module.cb_updata_persisted_object = this.updata_persisted_object.bind(this);
        this.hub_call_dbproxy_module.cb_remove_object = this.remove_object.bind(this);
        this.hub_call_dbproxy_module.cb_get_object_count = this.get_object_count.bind(this);
        this.hub_call_dbproxy_module.cb_get_object_info = this.get_object_info.bind(this);
    }

    public reg_hub(hub_name:string){
        log.getLogger().trace("hub:%s connected", hub_name);

		let _hubproxy = this._hubmng.reg_hub(this.hub_call_dbproxy_module.current_ch, hub_name);
		_hubproxy.reg_hub_sucess ();
    }
    
    public create_persisted_object(db:string, collection:string, object_info:string, callbackid:string){
        log.getLogger().trace("begin create_persisted_object");

        let _hubproxy = this._hubmng.get_hub(this.hub_call_dbproxy_module.current_ch);
        if (_hubproxy == null){
            log.getLogger().error("hubproxy is null");
            return;
        }

        this._mongodbproxy.save(db, collection, object_info, (is_save_sucess:boolean)=>{
            _hubproxy.ack_create_persisted_object(callbackid, is_save_sucess);
        })

        log.getLogger().trace("end create_persisted_object");
    }

    public updata_persisted_object(db:string, collection:string, query_json:string, object_info:string, callbackid:string){
        log.getLogger().trace("begin updata_persisted_object");

        let _hubproxy = this._hubmng.get_hub(this.hub_call_dbproxy_module.current_ch);
        if (_hubproxy == null){
            log.getLogger().error("hubproxy is null");
            return;
        }

        this._mongodbproxy.update(db, collection, query_json, object_info, (is_update_sucess:boolean)=>{
            _hubproxy.ack_updata_persisted_object(callbackid, is_update_sucess);
        });

        log.getLogger().trace("end updata_persisted_object");
    }

    public remove_object(db:string, collection:string, query_json:string, callbackid:string){
        log.getLogger().trace("begin remove_object");

        let _hubproxy = this._hubmng.get_hub(this.hub_call_dbproxy_module.current_ch);
        if (_hubproxy == null){
            log.getLogger().error("hubproxy is null");
            return;
        }

        this._mongodbproxy.remove(db, collection, query_json, (is_remove_sucess:boolean)=>{
            _hubproxy.ack_remove_object(callbackid, is_remove_sucess);
        });

        log.getLogger().trace("end remove_object");
    }

    public get_object_count(db:string, collection:string, query_json:string, callbackid:string){
        log.getLogger().trace("begin get_object_count");

        let _hubproxy = this._hubmng.get_hub(this.hub_call_dbproxy_module.current_ch);
        if (_hubproxy == null){
            log.getLogger().error("hubproxy is null");
            return;
        }

        this._mongodbproxy.count(db, collection, query_json, (count:number)=>{
            _hubproxy.ack_get_object_count(callbackid, count);
        });

        log.getLogger().trace("end get_object_count");
    }

    public get_object_info(db:string, collection:string, query_json:string, callbackid:string){
        log.getLogger().trace("begin get_object_info");

        let _hubproxy = this._hubmng.get_hub(this.hub_call_dbproxy_module.current_ch);
        if (_hubproxy == null){
            log.getLogger().error("hubproxy is null");
            return;
        }

        this._mongodbproxy.find(db, collection, query_json, (json_result:any[])=>{
            let count = 0;
            if (json_result.length == 0){
                _hubproxy.ack_get_object_info(callbackid, JSON.stringify([]));
            }
            else{
                let _datalist = [];
                for (let data of json_result)
                {
                    _datalist.push(data);

                    count++;
                    if (count >= 100){
                        _hubproxy.ack_get_object_info(callbackid, JSON.stringify(_datalist));

                        count = 0;
                        _datalist = [];
                    }
                }
                if (count > 0 && count < 100)
                {
                    _hubproxy.ack_get_object_info(callbackid, JSON.stringify(_datalist));
                }
            }
            _hubproxy.ack_get_object_info_end(callbackid);

        });

        log.getLogger().trace("end get_object_info");
    }
}

File: abelkhan_type_csharp_component_msgpack/component/dbproxy/hub_msg_handle.cs
using System;
using System.Collections;

namespace abelkhan
{
	public class hub_msg_handle
	{
        private dbevent _dbevent;
        private hubmanager _hubmanager;
        private mongodbproxy _mongodbproxy;
        private hub_call_dbproxy_module _module;

        public hub_msg_handle(modulemng modules, dbevent _dbevent_, hubmanager _hubmanager_, mongodbproxy _mongodbproxy_)
		{
            _dbevent = _dbevent_;
            _hubmanager = _hubmanager_;
			_mongodbproxy = _mongodbproxy_;

            _module = new hub_call_dbproxy_module(modules);
            _module.onreg_hub += reg_hub;
            _module.oncreate_persisted_object += create_persisted_object;
            _module.onupdata_persisted_object += updata_persisted_object;
            _module.onremove_object += remove_object;
            _module.onget_object_count += get_object_count;
            _module.onget_object_info += get_object_info;
            _module.onget_object_infoex += get_object_infoex;
        }

		public void reg_hub(string hub_name)
		{
            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "hub {0} connected", hub_name);

			hubproxy _hubproxy = _hubmanager.reg_hub(_module.current_ch, hub_name);
			_hubproxy.reg_hub_sucess ();
		}

		public void create_persisted_object(string db, string collection, string object_info, string callbackid)
		{
            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "begin create_persisted_object");

            hubproxy _hubproxy = _hubmanager.get_hub(_module.current_ch);
            if (_hubproxy == null)
            {
                log.error(new System.Diagnostics.StackFrame(true), timerservice.Tick, "hubproxy is null");
                return;
            }

            _dbevent.push_create_event(new create_event(_mongodbproxy, _hubproxy, db, collection, object_info, callbackid));

            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "end create_persisted_object");
        }

		public void updata_persisted_object(string db, string collection, string query_json, string object_info, string callbackid)
		{
            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "begin updata_persisted_object");

            hubproxy _hubproxy = _hubmanager.get_hub(_module.current_ch);
            if (_hubproxy == null)
            {
                log.error(new System.Diagnostics.StackFrame(true), timerservice.Tick, "hubproxy is null");
                return;
            }

            _dbevent.push_updata_event(new update_event(_mongodbproxy, _hubproxy, db, collection, query_json, object_info, callbackid));

            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "end updata_persisted_object");
        }

        public void remove_object(string db, string collection, string query_json, string callbackid)
        {
            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "begin remove_object");

            hubproxy _hubproxy = _hubmanager.get_hub(_module.current_ch);
            if (_hubproxy == null)
            {
                log.error(new System.Diagnostics.StackFrame(true), timerservice.Tick, "hubproxy is null");
                return;
            }

            _dbevent.push_remove_event(new remove_event(_mongodbproxy, _hubproxy, db, collection, query_json, callbackid));

            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "end remove_object");
        }

        public void get_object_count(string db, string collection, string query_json, string callbackid)
        {
            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "begin get_object_count");

            hubproxy _hubproxy = _hubmanager.get_hub(_module.current_ch);
            if (_hubproxy == null)
            {
                log.error(new System.Diagnostics.StackFrame(true), timerservice.Tick, "hubproxy is null");
                return;
            }

            _dbevent.push_count_event(new count_event(_mongodbproxy, _hubproxy, db, collection, query_json, callbackid));

            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "end get_object_count");
        }

		public void get_object_info(string db, string collection, string query_json, string callbackid)
        {
            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "begin get_object_info");

            hubproxy _hubproxy = _hubmanager.get_hub(_module.current_ch);
            if (_hubproxy == null)
            {
                log.error(new System.Diagnostics.StackFrame(true), timerservice.Tick, "hubproxy is null");
                return;
            }

            _dbevent.push_find_event(new find_event(_mongodbproxy, _hubproxy, db, collection, query_json, callbackid));

            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "end get_object_info");
        }

        public void get_object_infoex(string db, string collection, string query_json, int skip, int limit, string callbackid)
        {
            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "begin get_object_info");

            hubproxy _hubproxy = _hubmanager.get_hub(_module.current_ch);
            if (_hubproxy == null)
            {
                log.error(new System.Diagnostics.StackFrame(true), timerservice.Tick, "hubproxy is null");
                return;
            }

            _dbevent.push_findex_event(new findex_event(_mongodbproxy, _hubproxy, db, collection, query_json, skip, limit, callbackid));

            log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "end get_object_info");
        }
    }
}

